### ISSUE TITLE
[cloud_firestore_web] Firestore web calling Platform._operatingSystem resulting in crash

### ISSUE THREAD

**Issue by erparker** [FirebaseExtended/flutterfire#1998]
I started getting this error when running my flutter web app with cloud_firestore_web. 
I've been using cloud_firestore_web since before it was merged, and am using the new merged version, and didn't have any problems with it until today. I didn't change anything in my code, just ran `flutter upgrade` and started getting this issue

I'm using this in my pubspec

```
cloud_firestore: ^0.13.2+1
```

I'm building the app to Chrome using Android Studio. When I run the app I get a blank Chrome page and this in the javascript console:

```
Error: Unsupported operation: Platform._operatingSystem
    at Object.throw_ [as throw] (errors.dart:196)
    at Function._operatingSystem (io_patch.dart:241)
    at Function.get operatingSystem [as operatingSystem] (platform_impl.dart:62)
    at get _operatingSystem (platform.dart:73)
    at Function.desc.get [as _operatingSystem] (utils.dart:77)
    at get isIOS (platform.dart:141)
    at Function.desc.get [as isIOS] (utils.dart:77)
    at get firebaseDefaultAppName (default_app_name_io.dart:9)
    at Object.desc.get [as firebaseDefaultAppName] (utils.dart:77)
    at get defaultAppName (firebase_app.dart:19)
    at Function.desc.get [as defaultAppName] (utils.dart:77)
    at get instance (firebase_app.dart:41)
    at Function.desc.get [as instance] (utils.dart:77)
    at new cloud_firestore_web.FirestoreWeb.new (cloud_firestore_web.dart:33)
    at Function.registerWith (cloud_firestore_web.dart:26)
    at Object.registerPlugins (generated_plugin_registrant.dart:20)
    at main$ (web_entrypoint.dart:6)
    at main$.next (<anonymous>)
```

It sort of looks like instantiating a new cloud_firestore_web will call out to see what the operating system is, which is not allowed on flutter web. 

here's `flutter --version`
```
Flutter 1.15.4-pre.48 • channel master • https://github.com/flutter/flutter.git
Framework • revision 8abf54f5e3 (16 hours ago) • 2020-02-12 20:45:36 -0800
Engine • revision e0ebaea590
Tools • Dart 2.8.0 (build 2.8.0-dev.9.0 e4c39721c4)
```

flutter doctor -v:
```
[✓] Flutter (Channel master, v1.15.4-pre.48, on Mac OS X 10.14.6 18G84, locale en-US)
    • Flutter version 1.15.4-pre.48 at /Users/-/development/flutter
    • Framework revision 8abf54f5e3 (18 hours ago), 2020-02-12 20:45:36 -0800
    • Engine revision e0ebaea590
    • Dart version 2.8.0 (build 2.8.0-dev.9.0 e4c39721c4)

 
[!] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/-/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses

[✓] Xcode - develop for iOS and macOS (Xcode 11.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.3.1, Build version 11C504
    • CocoaPods version 1.6.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 43.0.1
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] VS Code (version 1.32.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 2.24.0

[✓] Connected device (3 available)
    • macOS      • macOS      • darwin-x64     • Mac OS X 10.14.6 18G84
    • Chrome     • chrome     • web-javascript • Google Chrome 80.0.3987.100
    • Web Server • web-server • web-javascript • Flutter Tools

! Doctor found issues in 1 category.
```

Any ideas?

**Comment by lfwells** [comment]
Confirmed I also have this error as of today. Was literally working for me yesterday, no changes in between except for switching from flutter dev branch back to flutter master branch (due to some issues there). About to test going back to dev branch to see if that fixes things but I doubt. 

To me it looks like the import 
`import 'default_app_name.dart' if (dart.library.io) 'default_app_name_io.dart';` is going to the _io version, which in turn calls Platform which I think isn't supported on web? the default_app_name.dart does NOT use Platform, and shouldn't crash as far as I can tell.

`errors.dart:147 Uncaught (in promise) Error: Unsupported operation: Platform._operatingSystem
    at Object.throw_ [as throw] (errors.dart:196)
    at Function._operatingSystem (io_patch.dart:241)
    at Function.get operatingSystem [as operatingSystem] (platform_impl.dart:62)
    at get _operatingSystem (platform.dart:73)
    at Function.desc.get [as _operatingSystem] (utils.dart:77)
    at get isIOS (platform.dart:141)
    at Function.desc.get [as isIOS] (utils.dart:77)
    at get firebaseDefaultAppName (default_app_name_io.dart:9)`

**Comment by erparker** [comment]
This change here looks relevant: https://github.com/FirebaseExtended/flutterfire/commit/c58abc0235330916206fed5d21bc3b7fad883f52

Looks like it should be stopping firestore from checking iOS if I am building to web, right? 

I'm also switching to dev just to see if that helps. 

My code isn't importing dart:io anywhere, which is almost irrelevant because execution stops before any of my other code is able to run

**Comment by lfwells** [comment]
Definitely looks relevant. This is where I start to get confused with plugin versions and patches etc, because *my* default_app_name.dart looks like this
`final String firebaseDefaultAppName =
    Platform.isIOS ? '__FIRAPP_DEFAULT' : '[DEFAULT]';`
whereas the commit you've referenced there appears to be adding the appropriate kWeb check which would prevent this error.

I'm personally only using `firebase_auth: ^0.15.4`, which in turn imports firebase_core . Is it using an old version of firebase_core? How do I get the fix represented in that commit (short term workaround would be to add the kIsWeb check to default_app_name.dart on your own copy.

**Comment by lfwells** [comment]
Also switching back to dev channel has fixed it for me for now, but I am keen to be on master to test out some other fixes they have done on there.

*confusion intensifies*

**Comment by erparker** [comment]
I forced importing firebase_core: 0.4.2+1 (which should contain that fix) and 0.4.4 but didn't fix the issue. 

I'm still in the middle of building the dev channel, I'll report back what I find

**Comment by lfwells** [comment]
okay things moving pretty fast here, I switch *back* to master to test some more (force import firebase_core etc), and now everything is working for me regardless of whether I force import or not. Had a bit of a look, this morning flutter doctor told me: `Flutter 1.15.4-pre.64 • channel master • https://github.com/flutter/flutter.git`, now it tells me `Flutter 1.15.4-pre.65`

@erparker will be interested to see how you went on dev channel, and then back to master channel without changing anything?

**Comment by erparker** [comment]
using the dev branch worked! It took 45 minutes to build after switching, which was odd to me, but everything was working normally again. 

I just switched back to master and did flutter upgrade, and it's once again taking quite a while to build the flutter tool. I'll build my add to chrome again once this flutter upgrade is done and I'll let you know. 

But yeah, this is a weird one. I don't know enough about the Dart VM to have a very educated guess here, but just because it's taking so long to build the tool, is it possible some super strange caching issue is causing this? Like, Flutter get confused when I built to web and just threw in the kitchen sink for some reason? No clue

**Comment by erparker** [comment]
Going back to the master branch resulted in a still broken web build. same error as before.

flutter --version 
```
Flutter 1.15.4-pre.65 • channel master • https://github.com/flutter/flutter.git
Framework • revision c725f107a1 (6 hours ago) • 2020-02-13 14:49:11 -0800
Engine • revision 0e8ed2715a
Tools • Dart 2.8.0 (build 2.8.0-dev.8.0 5244d99a5d)
```

I'm forcing firebase_core: 0.4.4 so it should have the fix in it mentioned earlier.

Just went back to the dev channel and it's working fine again. I am experiencing confusion.

**Comment by TimWhiting** [comment]
I think it is an issue with flutter & dart's conditional imports running web in debug mode, not the firebase_core. Take a look at [this](https://github.com/flutter/flutter/issues/50801)

**Comment by iapicca** [comment]
Hi @erparker 
could you please run

```console
flutter channel dev
flutter upgrade --force
```

if the issue persist
can you please provide your `flutter run --verbose` 
Thank you

**Comment by Salakar** [comment]
Hey, as part of our on-going work for #2582, this has been resolved in our Firebase Firestore rework (#2913) - which has now been merged into master. We'll look at publishing some prereleases in the next few days. Thank you
